imports/src/parser/import_parser.rs

```diff
@@ -0,0 +1,119 @@
+pub struct ImportParser {
+    imports: HashMap<String, Program>,
+    core_packages: Vec<Package>,
+}
+
+impl ImportParser {
+    ///
+    /// Creates a new empty `ImportParser`.
+    ///
+    pub fn new() -> Self {
+        Self {
+            imports: HashMap::new(),
+            core_packages: vec![],
+        }
+    }
+
+    ///
+    /// Inserts a (file name -> program) pair into the `ImportParser`.
+    ///
+    /// It is okay if the imported program is already present since importing multiple symbols from
+    /// the same file is allowed.
+    ///
+    pub(crate) fn insert_import(&mut self, file_name: String, program: Program) {
+        // Insert the imported program.
+        let _program = self.imports.insert(file_name.clone(), program);
+    }
+
+    ///
+    /// Inserts a core package into the `ImportParser`.
+    ///
+    /// If the vector did not have this file_name present, `Ok()` is returned.
+    ///
+    /// If the vector did have this file_name present, a duplicate import error is thrown.
+    ///
+    pub(crate) fn insert_core_package(&mut self, package: &Package) -> Result<(), ImportParserError> {
+        // Check for duplicate core package name.
+        if self.core_packages.contains(package) {
+            return Err(ImportParserError::duplicate_core_package(package.name.clone()));
+        }
+
+        // Append the core package.
+        self.core_packages.push(package.clone());
+
+        Ok(())
+    }
+
+    ///
+    /// Returns a reference to the program corresponding to the file name.
+    ///
+    pub fn get_import(&self, file_name: &str) -> Option<&Program> {
+        self.imports.get(file_name)
+    }
+
+    ///
+    /// Returns a reference to the core package corresponding to the given package.
+    ///
+    pub fn get_core_package(&self, package: &Package) -> Option<&Package> {
+        self.core_packages()
+            .iter()
+            .find(|core_package| core_package.eq(&package))
+    }
+
+    ///
+    /// Returns a reference to the vector of core packages.
+    ///
+    pub fn core_packages(&self) -> &Vec<Package> {
+        &self.core_packages
+    }
+
+    ///
+    /// Returns a new `ImportParser` from a given `Program`.
+    ///
+    /// For every import statement in the program:
+    ///     1. Check if the imported package exists.
+    ///     2. Create the typed syntax tree for the imported package.
+    ///     3. Insert the typed syntax tree into the `ImportParser`
+    ///
+    pub fn parse(program: &Program) -> Result<Self, ImportParserError> {
+        let mut imports = Self::new();
+
+        // Find all imports relative to current directory.
+        let path = current_dir().map_err(|error| ImportParserError::current_directory_error(error))?;
+
+        // Parse each import statement.
+        program
+            .imports
+            .iter()
+            .map(|import| imports.parse_package(path.clone(), &import.package))
+            .collect::<Result<Vec<()>, ImportParserError>>()?;
```

Comment:
if the resulting `Vec` doesn't matter, a plain `for` loop should be used instead, so that there are no allocations involved in the happy path